##the follwing code compares double-depletion with control cell lines, adjust the bam files for SON-depletion only or SRRM2-depletion only

#load the needed libraries
library(Rsamtools)
library(GenomicAlignments)
library(data.table)
library(GenomicFeatures)
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(regioneR)
library(ggrepel)


#input bam files
control_rep1_bam <- 'path/to/bam/file'
control_rep2_bam <- 'path/to/bam/file'
control_rep3_bam <- 'path/to/bam/file'

double_depl_rep1_bam <- 'path/to/bam/file'
double_depl_rep2_bam <- 'path/to/bam/file'
double_depl_rep3_bam <- 'path/to/bam/file'



# specify which nonmandatory fields to include
# (get full list by running scanBamWhat())
field_mapq <- c("mapq") 
param <- ScanBamParam(what=field_mapq)

# load the entire bam
control_rep1 <- readGAlignments(control_rep1_bam, param = param)
control_rep2 <- readGAlignments(control_rep2_bam, param = param)
control_rep3 <- readGAlignments(control_rep3_bam, param = param)

double_depl_rep1 <- readGAlignments(double_depl_rep1_bam, param = param)
double_depl_rep2 <- readGAlignments(double_depl_rep2_bam, param = param)
double_depl_rep3 <- readGAlignments(double_depl_rep3_bam, param = param)


# keep only unspliced reads
control_rep1 <- control_rep1[!grepl('N', cigar(control_rep1))] #N stands for "skipped region from the reference" 
control_rep2 <- control_rep2[!grepl('N', cigar(control_rep2))]
control_rep3 <- control_rep3[!grepl('N', cigar(control_rep3))]
double_depl_rep1 <- double_depl_rep1[!grepl('N', cigar(double_depl_rep1))]
double_depl_rep2 <- double_depl_rep2[!grepl('N', cigar(double_depl_rep2))]
double_depl_rep3 <- double_depl_rep3[!grepl('N', cigar(double_depl_rep3))]

#read bins file
bins <- read.table("../../shared/hg38_100kbins.saf", header = TRUE, sep="\t",stringsAsFactors=FALSE, quote="")
bins<- bins[, 1:4]
colnames(bins) <- c("name", "chrom", "chromStart", "chromEnd") #minimal bed file is ("chrom", "chromStart", "chromEnd")
bins<- bins[, c("chrom", "chromStart", "chromEnd", "name")] #change the order of the columns
rownames(bins) <- bins$name
bins.gr <- toGRanges(bins) #make it a GRanges object

# count reads
bins_control_rep1 <- countOverlaps(bins.gr, control_rep1, type = 'any', ignore.strand = TRUE)
bins_control_rep2 <- countOverlaps(bins.gr, control_rep2, type = 'any', ignore.strand = TRUE)
bins_control_rep3 <- countOverlaps(bins.gr, control_rep3, type = 'any', ignore.strand = TRUE)
bins_double_depl_rep1 <- countOverlaps(bins.gr, double_depl_rep1, type = 'any', ignore.strand = TRUE)
bins_double_depl_rep2 <- countOverlaps(bins.gr, double_depl_rep2, type = 'any', ignore.strand = TRUE)
bins_double_depl_rep3 <- countOverlaps(bins.gr, double_depl_rep3, type = 'any', ignore.strand = TRUE)



# get a table from the count vector

bins_control_rep1 <- data.table(bins_id = names(bins_control_rep1),
                             count   = unname(bins_control_rep1))
bins_control_rep2 <- data.table(bins_id = names(bins_control_rep2),
                                  count   = unname(bins_control_rep2))
bins_control_rep3 <- data.table(bins_id = names(bins_control_rep3),
                                    count   = unname(bins_control_rep3))
bins_double_depl_rep1 <- data.table(bins_id = names(bins_double_depl_rep1),
                                  count   = unname(bins_double_depl_rep1))
bins_double_depl_rep2 <- data.table(bins_id = names(bins_double_depl_rep2),
                                  count   = unname(bins_double_depl_rep3))
bins_double_depl_rep3 <- data.table(bins_id = names(bins_double_depl_rep2),
                                     count   = unname(bins_double_depl_rep3))


#save files
saveRDS(bins_control_rep1, file="../data/bins_control_rep1.Rda")
saveRDS(bins_control_rep2, file="../data/bins_control_rep2.Rda")
saveRDS(bins_control_rep3, file="../data/bins_control_rep3.Rda")
saveRDS(bins_double_depl_rep1, file="../data/bins_double_depl_rep1.Rda")
saveRDS(bins_double_depl_rep2, file="../data/bins_double_depl_rep2.Rda")
saveRDS(bins_double_depl_rep3, file="../data/bins_double_depl_rep3.Rda")


#load files (available in the data folder)
bins_control_rep1 <- readRDS(file="../data/bins_control_rep1.Rda")
bins_control_rep2 <- readRDS(file="../data/bins_control_rep2.Rda")
bins_control_rep3 <- readRDS(file="../data/bins_control_rep3.Rda")
bins_double_depl_rep1 <- readRDS(file="../data/bins_double_depl_rep1.Rda")
bins_double_depl_rep2 <- readRDS(file="../data/bins_double_depl_rep2.Rda")
bins_double_depl_rep3 <- readRDS(file="../data/bins_double_depl_rep3.Rda")



# build a common matrix where rows are the bins files
bins_control_rep1 <- bins_control_rep1[, 1:2]
colnames(bins_control_rep1) <- c("bins_id", "control_rep1") 
bins_control_rep2 <- bins_control_rep2[, 1:2]
colnames(bins_control_rep2) <- c("bins_id", "control_rep2") 
bins_control_rep3 <- bins_control_rep3[, 1:2]
colnames(bins_control_rep3) <- c("bins_id", "control_rep3") 

bins_double_depl_rep1 <- bins_double_depl_rep1[, 1:2]
colnames(bins_double_depl_rep1) <- c("bins_id", "double_depl_rep1") 
bins_double_depl_rep2 <- bins_double_depl_rep2[, 1:2]
colnames(bins_double_depl_rep2) <- c("bins_id", "double_depl_rep2") 
bins_double_depl_rep3 <- bins_double_depl_rep3[, 1:2]
colnames(bins_double_depl_rep3) <- c("bins_id", "double_depl_rep3") 

bins.counts.merged <- merge(bins_control_rep1, bins_control_rep2, by = "bins_id")
bins.counts.merged <- merge(bins.counts.merged, bins_control_rep3, by = "bins_id")
bins.counts.merged <- merge(bins.counts.merged, bins_double_depl_rep1, by = "bins_id")
bins.counts.merged <- merge(bins.counts.merged, bins_double_depl_rep2, by = "bins_id")
bins.counts.merged <- merge(bins.counts.merged, bins_double_depl_rep3, by = "bins_id")

bins.counts.merged <- bins.counts.merged %>% remove_rownames %>% column_to_rownames(var="bins_id")
saveRDS(bins.counts.merged, file="../data/double_bins.counts.merged.Rda")


## DESEQ2

#create sheet with conditions 
coldata <- data.table(
  rows = c("control_rep1", "control_rep2", "control_rep3", "double_depl_rep1", "double_depl_rep2","double_depl_rep3"),
  condition = c("control", "control","control", "double_depl", "double_depl", "double_depl"))
coldata <- coldata %>% remove_rownames %>% column_to_rownames(var="rows")

#run DE analysis
dds <- DESeqDataSetFromMatrix(countData = bins.counts.merged,
                              colData = coldata,
                              design = ~ condition)

keep <- rowSums(counts(dds)) >= 10 #keep if at least 10 reads in tot 
dds <- dds[keep,] 

dds <- DESeq(dds)

res <- results(dds)
res

resultsNames(dds) #write this as coefficient


# schrink
res_LFC <- lfcShrink(dds, coef="condition_double_depl_vs_control", type="apeglm")
res_LFC


# order them by p-value and save

res_LFC <- res_LFC[order(res_LFC$pvalue),]

saveRDS(res_LFC, file="../data/bins_double_vs_control_res_LFC.Rda")

write.csv(as.data.frame(res_LFC), 
          file="../data/bins_double_vs_control_res_LFC.csv")





