{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8ffc17b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "716ce21b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; gene_type \"transc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2742729</th>\n",
       "      <td>KI270744.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>51009</td>\n",
       "      <td>51114</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278625.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2742730</th>\n",
       "      <td>KI270744.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>51009</td>\n",
       "      <td>51114</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278625.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2742731</th>\n",
       "      <td>KI270750.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>148668</td>\n",
       "      <td>148843</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277374.1\"; gene_type \"snRNA\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2742732</th>\n",
       "      <td>KI270750.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>148668</td>\n",
       "      <td>148843</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277374.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2742733</th>\n",
       "      <td>KI270750.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>148668</td>\n",
       "      <td>148843</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277374.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2742734 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                chr   source     feature   start     end score strand frame  \\\n",
       "0              chr1   HAVANA        gene   11869   14409     .      +     .   \n",
       "1              chr1   HAVANA  transcript   11869   14409     .      +     .   \n",
       "2              chr1   HAVANA        exon   11869   12227     .      +     .   \n",
       "3              chr1   HAVANA        exon   12613   12721     .      +     .   \n",
       "4              chr1   HAVANA        exon   13221   14409     .      +     .   \n",
       "...             ...      ...         ...     ...     ...   ...    ...   ...   \n",
       "2742729  KI270744.1  ENSEMBL  transcript   51009   51114     .      -     .   \n",
       "2742730  KI270744.1  ENSEMBL        exon   51009   51114     .      -     .   \n",
       "2742731  KI270750.1  ENSEMBL        gene  148668  148843     .      +     .   \n",
       "2742732  KI270750.1  ENSEMBL  transcript  148668  148843     .      +     .   \n",
       "2742733  KI270750.1  ENSEMBL        exon  148668  148843     .      +     .   \n",
       "\n",
       "                                                 attribute  \n",
       "0        gene_id \"ENSG00000223972.5\"; gene_type \"transc...  \n",
       "1        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "2        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "3        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "4        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "...                                                    ...  \n",
       "2742729  gene_id \"ENSG00000278625.1\"; transcript_id \"EN...  \n",
       "2742730  gene_id \"ENSG00000278625.1\"; transcript_id \"EN...  \n",
       "2742731  gene_id \"ENSG00000277374.1\"; gene_type \"snRNA\"...  \n",
       "2742732  gene_id \"ENSG00000277374.1\"; transcript_id \"EN...  \n",
       "2742733  gene_id \"ENSG00000277374.1\"; transcript_id \"EN...  \n",
       "\n",
       "[2742734 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read genes gtf file\n",
    "\n",
    "genes_path='path-to-your/genes.gtf'\n",
    "column_names = [\"chr\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "genes = pd.read_csv(genes_path, sep='\\t', comment='#', header=None, names=column_names)\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7db989cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>GeneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2742636</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>72411</td>\n",
       "      <td>74814</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000276017.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2742643</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>131494</td>\n",
       "      <td>137392</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000278817.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2742659</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>138082</td>\n",
       "      <td>161852</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2742728</th>\n",
       "      <td>KI270744.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>51009</td>\n",
       "      <td>51114</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278625.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2742731</th>\n",
       "      <td>KI270750.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>148668</td>\n",
       "      <td>148843</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000277374.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58780 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                chr feature   start     end strand             GeneID\n",
       "0              chr1    gene   11869   14409      +  ENSG00000223972.5\n",
       "12             chr1    gene   14404   29570      -  ENSG00000227232.5\n",
       "25             chr1    gene   17369   17436      -  ENSG00000278267.1\n",
       "28             chr1    gene   29554   31109      +  ENSG00000243485.5\n",
       "36             chr1    gene   30366   30503      +  ENSG00000284332.1\n",
       "...             ...     ...     ...     ...    ...                ...\n",
       "2742636  KI270734.1    gene   72411   74814      +  ENSG00000276017.1\n",
       "2742643  KI270734.1    gene  131494  137392      +  ENSG00000278817.1\n",
       "2742659  KI270734.1    gene  138082  161852      -  ENSG00000277196.4\n",
       "2742728  KI270744.1    gene   51009   51114      -  ENSG00000278625.1\n",
       "2742731  KI270750.1    gene  148668  148843      +  ENSG00000277374.1\n",
       "\n",
       "[58780 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#select only gene features and extract gene_id from attributes\n",
    "\n",
    "genes['GeneID'] = genes['attribute'].apply(lambda x: next((entry.split(' ')[1].strip('\"') for entry in x.split(';') if entry.strip().startswith('gene_id')), None))\n",
    "genes = genes[['chr', 'feature', 'start', 'end', 'strand', 'GeneID']]\n",
    "genes = genes[genes['feature'] == 'gene']\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1a9f101e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    58780\n",
       "Name: GeneID, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check for duplicate gene IDs\n",
    "\n",
    "genes['GeneID'].duplicated().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3795d84a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>GeneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741715</th>\n",
       "      <td>chrY</td>\n",
       "      <td>gene</td>\n",
       "      <td>57184101</td>\n",
       "      <td>57197337</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000124334.17_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741770</th>\n",
       "      <td>chrY</td>\n",
       "      <td>gene</td>\n",
       "      <td>57190738</td>\n",
       "      <td>57208756</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000270726.6_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741776</th>\n",
       "      <td>chrY</td>\n",
       "      <td>gene</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741780</th>\n",
       "      <td>chrY</td>\n",
       "      <td>gene</td>\n",
       "      <td>57207346</td>\n",
       "      <td>57212230</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741867</th>\n",
       "      <td>chrY</td>\n",
       "      <td>gene</td>\n",
       "      <td>57212184</td>\n",
       "      <td>57214397</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58684 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr feature     start       end strand                    GeneID\n",
       "0        chr1    gene     11869     14409      +         ENSG00000223972.5\n",
       "12       chr1    gene     14404     29570      -         ENSG00000227232.5\n",
       "25       chr1    gene     17369     17436      -         ENSG00000278267.1\n",
       "28       chr1    gene     29554     31109      +         ENSG00000243485.5\n",
       "36       chr1    gene     30366     30503      +         ENSG00000284332.1\n",
       "...       ...     ...       ...       ...    ...                       ...\n",
       "2741715  chrY    gene  57184101  57197337      +  ENSG00000124334.17_PAR_Y\n",
       "2741770  chrY    gene  57190738  57208756      +   ENSG00000270726.6_PAR_Y\n",
       "2741776  chrY    gene  57201143  57203357      -  ENSG00000185203.12_PAR_Y\n",
       "2741780  chrY    gene  57207346  57212230      +  ENSG00000182484.15_PAR_Y\n",
       "2741867  chrY    gene  57212184  57214397      -   ENSG00000227159.8_PAR_Y\n",
       "\n",
       "[58684 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#select only standard chromosomes\n",
    "\n",
    "chroms = ['chr'+str(x) for x in range(1,23)] + ['chrX', 'chrY']\n",
    "genes = genes[(genes.chr.isin(chroms))]\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b91dfa71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>GeneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741568</th>\n",
       "      <td>chrY</td>\n",
       "      <td>gene</td>\n",
       "      <td>26549425</td>\n",
       "      <td>26549743</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000224240.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741571</th>\n",
       "      <td>chrY</td>\n",
       "      <td>gene</td>\n",
       "      <td>26586642</td>\n",
       "      <td>26591601</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227629.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741576</th>\n",
       "      <td>chrY</td>\n",
       "      <td>gene</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000237917.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741591</th>\n",
       "      <td>chrY</td>\n",
       "      <td>gene</td>\n",
       "      <td>26626520</td>\n",
       "      <td>26627159</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000231514.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741594</th>\n",
       "      <td>chrY</td>\n",
       "      <td>gene</td>\n",
       "      <td>56855244</td>\n",
       "      <td>56855488</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000235857.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58639 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr feature     start       end strand             GeneID\n",
       "0        chr1    gene     11869     14409      +  ENSG00000223972.5\n",
       "12       chr1    gene     14404     29570      -  ENSG00000227232.5\n",
       "25       chr1    gene     17369     17436      -  ENSG00000278267.1\n",
       "28       chr1    gene     29554     31109      +  ENSG00000243485.5\n",
       "36       chr1    gene     30366     30503      +  ENSG00000284332.1\n",
       "...       ...     ...       ...       ...    ...                ...\n",
       "2741568  chrY    gene  26549425  26549743      +  ENSG00000224240.1\n",
       "2741571  chrY    gene  26586642  26591601      -  ENSG00000227629.1\n",
       "2741576  chrY    gene  26594851  26634652      -  ENSG00000237917.1\n",
       "2741591  chrY    gene  26626520  26627159      -  ENSG00000231514.1\n",
       "2741594  chrY    gene  56855244  56855488      +  ENSG00000235857.1\n",
       "\n",
       "[58639 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filter for Ensembl gene IDs only\n",
    "\n",
    "genes = genes[genes['GeneID'].str.contains(r'^ENSG\\d+\\.\\d+$')]\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2278820e",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes[['GeneID', 'chr', 'start', 'end', 'strand']].to_csv('genes.saf', index=False, header=False, sep='\\t') #this file is provided in the current directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a0b90f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load the filtered bam files output from snakepipes mRNA-seq pipeline\n",
    "\n",
    "bam = \"path-to-directory/filtered_bam\" #not provided here because too large"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19c7c40a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.6\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m8 BAM files  \u001b[0m \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mLM124.filtered.bam\u001b[0m \u001b[0m                              ||\n",
      "||                           \u001b[36mLM125.filtered.bam\u001b[0m \u001b[0m                              ||\n",
      "||                           \u001b[36mLM126.filtered.bam\u001b[0m \u001b[0m                              ||\n",
      "||                           \u001b[36mLM127.filtered.bam\u001b[0m \u001b[0m                              ||\n",
      "||                           \u001b[36mLM128.filtered.bam\u001b[0m \u001b[0m                              ||\n",
      "||                           \u001b[36mLM129.filtered.bam\u001b[0m \u001b[0m                              ||\n",
      "||                           \u001b[36mLM130.filtered.bam\u001b[0m \u001b[0m                              ||\n",
      "||                           \u001b[36mLM131.filtered.bam\u001b[0m \u001b[0m                              ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mall_featureCounts.txt\u001b[0m \u001b[0m                           ||\n",
      "||                 Summary : \u001b[36mall_featureCounts.txt.summary\u001b[0m \u001b[0m                   ||\n",
      "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mgenes.saf (SAF)\u001b[0m \u001b[0m                                 ||\n",
      "||      Dir for temp files : \u001b[36m/scratch/local2/martina\u001b[0m \u001b[0m                         ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m8\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||     Multiple alignments : \u001b[36mprimary alignment only\u001b[0m \u001b[0m                          ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file genes.saf \u001b[0m... \u001b[0m                                        ||\n",
      "||    Features : \u001b[36m58639\u001b[0m \u001b[0m                                                       ||\n",
      "||    Meta-features : \u001b[36m58639\u001b[0m \u001b[0m                                                  ||\n",
      "||    Chromosomes/contigs : \u001b[36m24\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file LM124.filtered.bam... \u001b[0m                                    ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
      "||    Total alignments : \u001b[36m45429013\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m33606007 (74.0%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.18 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file LM125.filtered.bam... \u001b[0m                                    ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
      "||    Total alignments : \u001b[36m36701319\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m26332700 (71.7%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.15 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file LM126.filtered.bam... \u001b[0m                                    ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
      "||    Total alignments : \u001b[36m35269392\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m26104881 (74.0%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.15 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file LM127.filtered.bam... \u001b[0m                                    ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
      "||    Total alignments : \u001b[36m49900353\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m36040074 (72.2%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.19 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file LM128.filtered.bam... \u001b[0m                                    ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
      "||    Total alignments : \u001b[36m43353978\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m30108425 (69.4%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.17 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file LM129.filtered.bam... \u001b[0m                                    ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
      "||    Total alignments : \u001b[36m30152199\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m21985248 (72.9%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.12 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file LM130.filtered.bam... \u001b[0m                                    ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
      "||    Total alignments : \u001b[36m38064162\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m27345050 (71.8%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.16 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file LM131.filtered.bam... \u001b[0m                                    ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
      "||    Total alignments : \u001b[36m37272825\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m26506481 (71.1%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.15 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"/project/bigdata_aktas/ \u001b[0m ||\n",
      "|| analysis/LM124-131_4sU_CSSc50/DE_on_introns/python/all_featureCounts.txt. \u001b[0m ||\n",
      "|| summary\" \u001b[0m                                                                  ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# see https:/bioconductor.org/packages/release/bioc/vignettes/Rsubread/inst/doc/SubreadUsersGuide.pdf\n",
    "# download from https:/rnnh.github.io/bioinfo-notebook/docs/featureCounts.html\n",
    "\n",
    "!featureCounts -F SAF \\\n",
    "--primary \\\n",
    "-p -B \\\n",
    "-Q 10 \\\n",
    "-s 0 \\\n",
    "-T 8 \\\n",
    "--tmpDir [path-to-directory-to-store-temporary-files] \\\n",
    "-a genes.saf \\\n",
    "-o [path-to-directory-to-store-output-files]/all_featureCounts.txt \\\n",
    "[path-to-directory]/filtered_bam/sample1.filtered.bam [path-to-directory]/filtered_bam/sample2.filtered.bam [path-to-directory]/filtered_bam/sampleN.filtered.bam\n",
    "\n",
    "# -F creates annotation index\n",
    "# --primary only primary alignments will be counted\n",
    "# -p is pairedEnd\n",
    "# -B is only fragments that have both ends successfully aligned will be considered for summarization\n",
    "# -Q minimum mapping quality score\n",
    "# -s 0 means un-stranded (KAPA kit isn't stranded enough)\n",
    "# -T number of threads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d23b636b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>control_1</th>\n",
       "      <th>control_2</th>\n",
       "      <th>SRRM2depl_1</th>\n",
       "      <th>SRRM2depl_2</th>\n",
       "      <th>SONdepl_1</th>\n",
       "      <th>SONdepl_2</th>\n",
       "      <th>doubledepl_1</th>\n",
       "      <th>doubledepl_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>149</td>\n",
       "      <td>103</td>\n",
       "      <td>86</td>\n",
       "      <td>88</td>\n",
       "      <td>91</td>\n",
       "      <td>53</td>\n",
       "      <td>73</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58634</th>\n",
       "      <td>ENSG00000224240.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58635</th>\n",
       "      <td>ENSG00000227629.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58636</th>\n",
       "      <td>ENSG00000237917.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58637</th>\n",
       "      <td>ENSG00000231514.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58638</th>\n",
       "      <td>ENSG00000235857.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58639 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Geneid  control_1  control_2  SRRM2depl_1  SRRM2depl_2  \\\n",
       "0      ENSG00000223972.5          1          1            0            1   \n",
       "1      ENSG00000227232.5        149        103           86           88   \n",
       "2      ENSG00000278267.1          0          0            0            0   \n",
       "3      ENSG00000243485.5          0          2            2            2   \n",
       "4      ENSG00000284332.1          0          0            0            0   \n",
       "...                  ...        ...        ...          ...          ...   \n",
       "58634  ENSG00000224240.1          0          0            0            0   \n",
       "58635  ENSG00000227629.1          0          0            0            0   \n",
       "58636  ENSG00000237917.1          0          0            0            4   \n",
       "58637  ENSG00000231514.1          0          0            0            0   \n",
       "58638  ENSG00000235857.1          0          0            0            0   \n",
       "\n",
       "       SONdepl_1  SONdepl_2  doubledepl_1  doubledepl_2  \n",
       "0              1          3             3             4  \n",
       "1             91         53            73            73  \n",
       "2              0          0             0             0  \n",
       "3              2          0             1             2  \n",
       "4              0          0             0             0  \n",
       "...          ...        ...           ...           ...  \n",
       "58634          0          0             0             0  \n",
       "58635          0          0             0             0  \n",
       "58636          0          0             0             0  \n",
       "58637          0          0             0             0  \n",
       "58638          0          0             0             0  \n",
       "\n",
       "[58639 rows x 9 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load the output files from the previous command\n",
    "\n",
    "counts = pd.read_table('../data/all_featureCounts.txt', comment='#') \n",
    "counts.columns = [\n",
    "    \"Geneid\", \n",
    "    \"Chr\", \n",
    "    \"Start\", \n",
    "    \"End\", \n",
    "    \"Strand\", \n",
    "    \"Length\", \n",
    "    \"control_1\", \n",
    "    \"control_2\", \n",
    "    \"SRRM2depl_1\", \n",
    "    \"SRRM2depl_2\", \n",
    "    \"SONdepl_1\", \n",
    "    \"SONdepl_2\", \n",
    "    \"doubledepl_1\", \n",
    "    \"doubledepl_2\", \n",
    "]\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8157cc2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save count tables for different comparisons\n",
    "\n",
    "counts_SON = counts[['Geneid',\"control_1\", \"control_2\", \"SONdepl_1\", \"SONdepl_2\"]] \n",
    "counts_SRRM2 = counts[['Geneid', \"control_1\", \"control_2\", \"SRRM2depl_1\", \"SRRM2depl_2\"]]\n",
    "counts_double = counts[['Geneid', \"control_1\", \"control_2\", \"doubledepl_1\", \"doubledepl_2\"]] \n",
    "counts_all = counts[['Geneid', \"control_1\", \"control_2\", \"SRRM2depl_1\", \"SRRM2depl_2\", \"SONdepl_1\", \"SONdepl_2\", \"doubledepl_1\", \"doubledepl_2\"]] \n",
    "\n",
    "counts_SON.to_csv(\"../data/SONdepl_featureCounts.txt\", sep='\\t', index=False)\n",
    "counts_SRRM2.to_csv(\"../data/SRRM2depl_featureCounts.txt\", sep='\\t', index=False)\n",
    "counts_double.to_csv(\"../data/Doubledepl_featureCounts.txt\", sep='\\t', index=False)\n",
    "counts_all.to_csv(\"../data/all_featureCounts.txt\", sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scipy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
